# Importing Required Libraries

In [84]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Loading Dataset

In [85]:
df = pd.read_excel('online_retail_II.xlsx')

## Understanding Dataset

In [86]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [87]:
df.shape

(525461, 8)

###  Dataset Overview

- Total Rows (Raw): 525,461
- Total Columns: 8

Columns:
- Invoice
- StockCode
- Description
- Quantity
- InvoiceDate
- Price
- CustomerID
- Country

The dataset contains transactional retail sales data across multiple countries.

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [89]:
df.describe()

Quantity                    InvoiceDate          Price  \
count  525461.000000                         525461  525461.000000   
mean       10.337667  2010-06-28 11:37:36.845017856       4.688834   
min     -9600.000000            2009-12-01 07:45:00  -53594.360000   
25%         1.000000            2010-03-21 12:20:00       1.250000   
50%         3.000000            2010-07-06 09:51:00       2.100000   
75%        10.000000            2010-10-15 12:45:00       4.210000   
max     19152.000000            2010-12-09 20:01:00   25111.090000   
std       107.424110                            NaN     146.126914   

         Customer ID  
count  417534.000000  
mean    15360.645478  
min     12346.000000  
25%     13983.000000  
50%     15311.000000  
75%     16799.000000  
max     18287.000000  
std      1680.811316

In [90]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [91]:
df['InvoiceDate'].dtype

dtype('<M8[ns]')

In [92]:
df.duplicated().sum()

np.int64(6865)

###  Initial Observations

- Dataset contains 525461 rows and 8 columns.
- No. of numeric columns: 3
- No. of categorical columns: 4
- Some missing values observed in Description and Customer_ID columns.
- Dataset contains 6865 duplicates.

##  Data Cleaning

### Creating copy of Original Dataset

In [93]:
dfcopy = df.copy()

In [94]:
dfcopy.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [95]:
## Cheaking duplicates
dfcopy.duplicated().sum()

np.int64(6865)

In [96]:
## Droping duplicates
dfcopy.drop_duplicates(keep = 'first',inplace = True)

In [97]:
dfcopy.shape

(518596, 8)

In [98]:
dfcopy['Invoice'].unique()

array([489434, 489435, 489436, ..., 538169, 538170, 538171],
      shape=(28816,), dtype=object)

In [99]:
dfcopy['StockCode'].unique()

array([85048, '79323P', '79323W', ..., 22935, 22933, 21120],
      shape=(4632,), dtype=object)

In [100]:
dfcopy.describe()

Quantity                    InvoiceDate          Price  \
count  518596.000000                         518596  518596.000000   
mean       10.438152  2010-06-28 06:43:01.573016320       4.716176   
min     -9600.000000            2009-12-01 07:45:00  -53594.360000   
25%         1.000000            2010-03-21 11:07:00       1.250000   
50%         3.000000            2010-07-05 15:42:00       2.100000   
75%        10.000000            2010-10-15 11:40:00       4.210000   
max     19152.000000            2010-12-09 20:01:00   25111.090000   
std       108.126593                            NaN     147.090460   

         Customer ID  
count  410763.000000  
mean    15353.621857  
min     12346.000000  
25%     13979.000000  
50%     15311.000000  
75%     16797.000000  
max     18287.000000  
std      1681.657665

In [101]:
dfcopy.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107833
Country             0
dtype: int64

In [102]:
negative_qty_count = (dfcopy['Quantity']<0).sum()
print('Negative Quantity Rows:',negative_qty_count)

total_rows = len(dfcopy)
print('Total Rows:', total_rows)

pct_of_negative_qty_count = (negative_qty_count/total_rows)*100
print('Percentage of negative qty value:' ,pct_of_negative_qty_count)

Negative Quantity Rows: 12302
Total Rows: 518596
Percentage of negative qty value: 2.3721741008415025


#### Observation 
- Dataset has 12302 negative Quantity rows.
- In parcentage nearly 2.38%

In [103]:
dfcopy['Invoice'].unique()

array([489434, 489435, 489436, ..., 538169, 538170, 538171],
      shape=(28816,), dtype=object)

In [104]:
# count_c = df.loc[df['Invoice'].astype(str).str.startswith('C'),'Invoice'].unique()
# print('start with C:',count_c)

count_c = dfcopy['Invoice'].astype(str).str.startswith('C').sum()
print('start with C:',count_c)

start with C: 10182


In [105]:
negative_qty = dfcopy[dfcopy['Quantity']<0]
check_c = negative_qty['Invoice'].astype(str).str.startswith('C')



In [106]:
all_c = check_c.all()
print('Do all negative quantities belong to C invoices?', all_c)

Do all negative quantities belong to C invoices? False


In [107]:
ngtv_price = (dfcopy['Price']<0).sum()
print('Price that are in negative:',ngtv_price)

zero_price = (dfcopy['Price']==0).sum()
print('Number are price that are zero:',zero_price)

Price that are in negative: 3
Number are price that are zero: 3681


In [108]:
dfcopy['Price'].describe()

count    518596.000000
mean          4.716176
std         147.090460
min      -53594.360000
25%           1.250000
50%           2.100000
75%           4.210000
max       25111.090000
Name: Price, dtype: float64

In [109]:
dfcopy['Price'].quantile(0.99)

np.float64(20.0)

In [110]:
dfcopy.sort_values(by = 'Price',ascending = False ).head(10)

Invoice     StockCode   Description  Quantity         InvoiceDate  \
241827   512771             M        Manual         1 2010-06-17 16:53:00   
241824  C512770             M        Manual        -1 2010-06-17 16:52:00   
320581  C520667  BANK CHARGES  Bank Charges        -1 2010-08-27 13:42:00   
517955   537632     AMAZONFEE    AMAZON FEE         1 2010-12-07 15:08:00   
517953  C537630     AMAZONFEE    AMAZON FEE        -1 2010-12-07 15:04:00   
519294  C537651     AMAZONFEE    AMAZON FEE        -1 2010-12-07 15:49:00   
519170  C537644     AMAZONFEE    AMAZON FEE        -1 2010-12-07 15:34:00   
135012  C502262             M        Manual        -1 2010-03-23 15:20:00   
135015   502265             M        Manual         1 2010-03-23 15:28:00   
135014  C502264             M        Manual        -1 2010-03-23 15:24:00   

           Price  Customer ID         Country  
241827  25111.09          NaN  United Kingdom  
241824  25111.09      17399.0  United Kingdom  
320581  18910.69          NaN  United Kingdom  
517955  13541.33          NaN  United Kingdom  
517953  13541.33          NaN  United Kingdom  
519294  13541.33          NaN  United Kingdom  
519170  13474.79          NaN  United Kingdom  
135012  10953.50      12918.0  United Kingdom  
135015  10953.50          NaN  United Kingdom  
135014  10953.50      12918.0  United Kingdom

In [111]:
# step 1: Define keywords 
adjustment_keywords = ['manual','amazon','bank']
#step 2: create pattern 
pattern = '|'.join(adjustment_keywords)
#step 3: Filter out adjustment rows
df_no_adjustments = dfcopy[~dfcopy['Description'].str.contains(pattern, case = False, na=False)]

In [112]:
df_no_adjustments['Price'].max()

5117.03

In [113]:
df_no_adjustments.sort_values(by='Price', ascending = False).head(10)

Invoice StockCode                          Description  Quantity  \
71077    495798    ADJUST  Adjustment by john on 26/01/2010 17         1   
450818  C532255         D                             Discount        -1   
192196   507637     84016           FLAG OF ST GEORGE CAR FLAG         1   
62735    494918       DOT                       DOTCOM POSTAGE         1   
516862   537434       DOT                       DOTCOM POSTAGE         1   
95554   C498420         D                             Discount        -1   
36616    492414       DOT                       DOTCOM POSTAGE         1   
514160   537240       DOT                       DOTCOM POSTAGE         1   
9701     490149       DOT                       DOTCOM POSTAGE         1   
509103   536876       DOT                       DOTCOM POSTAGE         1   

               InvoiceDate    Price  Customer ID         Country  
71077  2010-01-26 17:25:00  5117.03          NaN  United Kingdom  
450818 2010-11-11 12:50:00  1269.51      14088.0  United Kingdom  
192196 2010-05-10 14:55:00  1157.15          NaN  United Kingdom  
62735  2010-01-19 17:49:00  1081.70          NaN  United Kingdom  
516862 2010-12-06 16:57:00   950.99          NaN  United Kingdom  
95554  2010-02-18 16:25:00   947.08      14646.0     Netherlands  
36616  2009-12-16 17:01:00   945.56          NaN  United Kingdom  
514160 2010-12-06 10:08:00   940.87          NaN  United Kingdom  
9701   2009-12-04 09:43:00   922.05          NaN  United Kingdom  
509103 2010-12-03 11:36:00   887.52          NaN  United Kingdom

In [114]:
# Step 1: Define non-product keywords
non_product_keywords = ['manual', 'bank', 'amazon', 'adjustment', 'postage', 'discount']

pattern = '|'.join(non_product_keywords)

# Step 2: Create product sales dataset
dfcopy_sales = dfcopy[
    (dfcopy['Quantity'] > 0) &
    (dfcopy['Price'] > 0) &
    (~dfcopy['Description'].str.contains(pattern, case=False, na=False))
]


In [115]:
dfcopy_sales.shape

(500654, 8)

In [116]:
dfcopy_sales['Price'].max()

1157.15

In [117]:
dfcopy_sales['Price'].quantile(0.99)

np.float64(16.98)

In [118]:
dfcopy_sales['Quantity'].describe()


count    500654.000000
mean         11.565474
std          87.685129
min           1.000000
25%           1.000000
50%           3.000000
75%          12.000000
max       19152.000000
Name: Quantity, dtype: float64

In [119]:
dfcopy_sales['Quantity'].quantile(0.99)

np.float64(120.0)

In [120]:
dfcopy_sales.sort_values(by = 'Quantity', ascending = False).head(10)

Invoice StockCode                         Description  Quantity  \
90857   497946     37410  BLACK AND WHITE PAISLEY FLOWER MUG     19152   
127168  501534     21091         SET/6 WOODLAND PAPER PLATES     12960   
127166  501534     21099         SET/6 STRAWBERRY PAPER CUPS     12960   
127169  501534     21085           SET/6 WOODLAND PAPER CUPS     12744   
127167  501534     21092       SET/6 STRAWBERRY PAPER PLATES     12480   
135027  502269     21984    PACK OF 12 PINK PAISLEY TISSUES      10000   
135030  502269     21981        PACK OF 12 WOODLAND TISSUES      10000   
135028  502269     21982            PACK OF 12 SUKI TISSUES      10000   
135029  502269     21980      PACK OF 12 RED SPOTTY TISSUES      10000   
93677   498152     85220     SMALL FAIRY CAKE FRIDGE MAGNETS      9456   

               InvoiceDate  Price  Customer ID         Country  
90857  2010-02-15 11:57:00   0.10      13902.0         Denmark  
127168 2010-03-17 13:09:00   0.10      13902.0         Denmark  
127166 2010-03-17 13:09:00   0.10      13902.0         Denmark  
127169 2010-03-17 13:09:00   0.10      13902.0         Denmark  
127167 2010-03-17 13:09:00   0.10      13902.0         Denmark  
135027 2010-03-23 15:36:00   0.25      17940.0  United Kingdom  
135030 2010-03-23 15:36:00   0.25      17940.0  United Kingdom  
135028 2010-03-23 15:36:00   0.25      17940.0  United Kingdom  
135029 2010-03-23 15:36:00   0.25      17940.0  United Kingdom  
93677  2010-02-17 10:51:00   0.30      13902.0         Denmark

### Creating Revenue Column

In [121]:
dfcopy_sales['Revenue'] = dfcopy_sales['Quantity']*dfcopy_sales['Price']

C:\Users\rahul\AppData\Local\Temp\ipykernel_16048\3739093656.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfcopy_sales['Revenue'] = dfcopy_sales['Quantity']*dfcopy_sales['Price']


In [122]:
countc = dfcopy_sales['Invoice'].astype(str).str.startswith('C').sum()
print('start with C:',countc)

start with C: 0


In [123]:
dfcopy_sales.shape

(500654, 9)

## Data Cleaning & Preprocessing

####  Steps Performed:

- Removed duplicate rows (6,865 duplicates removed)
- Filtered only product sales:
    - Quantity > 0
    - Price > 0
- Removed cancellation invoices (Invoice starting with 'C')
- Removed rows with service/adjustment descriptions:
    - Manual
    - Discount
    - Amazon Fee
    - Adjustment
- Handled outliers using 99th percentile filtering
- Created Revenue column:
    
    Revenue = Quantity * Price

#### Final Clean Dataset:

- Total Rows: 500,654
- No duplicates
- Only valid product transactions

In [124]:
dfcopy_sales['Revenue'].sum()

np.float64(9785571.004)

In [125]:
top_10_revenue_descri = (dfcopy_sales.groupby('Description')['Revenue'].sum().sort_values(ascending = False).head(10))

In [126]:
top_10_revenue_descri

Description
REGENCY CAKESTAND 3 TIER               169912.76
WHITE HANGING HEART T-LIGHT HOLDER     160345.63
ASSORTED COLOUR BIRD ORNAMENT           72890.19
PAPER CHAIN KIT 50'S CHRISTMAS          58127.30
JUMBO BAG RED RETROSPOT                 56480.46
PARTY BUNTING                           49664.12
ROTATING SILVER ANGELS T-LIGHT HLDR     47954.49
EDWARDIAN PARASOL NATURAL               36925.26
JUMBO BAG STRAWBERRY                    36732.16
VINTAGE UNION JACK BUNTING              36023.71
Name: Revenue, dtype: float64

In [127]:
top_10_quantity_descri = (dfcopy_sales.groupby('Description')['Quantity'].sum().sort_values(ascending = False).head(10))
top_10_quantity_descri

Description
WHITE HANGING HEART T-LIGHT HOLDER    58691
WORLD WAR 2 GLIDERS ASSTD DESIGNS     54947
BROCADE RING PURSE                    48374
PACK OF 72 RETRO SPOT CAKE CASES      46728
ASSORTED COLOUR BIRD ORNAMENT         45228
60 TEATIME FAIRY CAKE CASES           36348
PACK OF 60 PINK PAISLEY CAKE CASES    31805
JUMBO BAG RED RETROSPOT               30746
SMALL POPCORN HOLDER                  29773
STRAWBERRY CERAMIC TRINKET BOX        27059
Name: Quantity, dtype: int64

In [128]:
top_10_avgprice_descri = (dfcopy_sales.groupby('Description')['Price'].mean().sort_values(ascending = False).head(10))
top_10_avgprice_descri

Description
VINTAGE BLUE KITCHEN CABINET           295.000000
VINTAGE RED KITCHEN CABINET            295.000000
RUSTIC  SEVENTEEN DRAWER SIDEBOARD     161.000000
GIANT SEVENTEEN DRAWER SIDEBOARD       158.333333
REGENCY MIRROR WITH SHUTTERS           153.000000
FRENCH STYLE WALL DRESSER              116.923077
CHEST NATURAL WOOD 20 DRAWERS          116.000000
ANT WHITE SWEETHEART TABLE W 3 DRAW    106.666667
VINTAGE POST OFFICE CABINET             77.626667
SAMPLES                                 73.800000
Name: Price, dtype: float64

In [129]:
dfcopy_sales.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  Revenue  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom     83.4  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom    100.8  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom     30.0

In [130]:
dfcopy_sales['YearMonth'] = dfcopy_sales['InvoiceDate'].dt.strftime('%Y-%m')
dfcopy_sales['YearMonth'].unique()

C:\Users\rahul\AppData\Local\Temp\ipykernel_16048\1274936117.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfcopy_sales['YearMonth'] = dfcopy_sales['InvoiceDate'].dt.strftime('%Y-%m')


array(['2009-12', '2010-01', '2010-02', '2010-03', '2010-04', '2010-05',
       '2010-06', '2010-07', '2010-08', '2010-09', '2010-10', '2010-11',
       '2010-12'], dtype=object)

In [131]:
monthly_revenue = (dfcopy_sales.groupby('YearMonth')['Revenue'].sum().sort_values(ascending = False))
monthly_revenue.head()

YearMonth
2010-11    1426085.552
2010-10    1091642.650
2010-09     866815.201
2009-12     797577.130
2010-03     761444.501
Name: Revenue, dtype: float64

In [132]:
Country_revenue = dfcopy_sales.groupby('Country')['Revenue'].sum().sort_values(ascending = False)
Country_revenue.head()

Country
United Kingdom    8431457.713
EIRE               361996.940
Netherlands        265635.270
Germany            182409.631
France             132135.310
Name: Revenue, dtype: float64

In [133]:
Top_revenue_customerId = (dfcopy_sales.groupby('Customer ID').agg({'Revenue': 'sum','Quantity': 'sum', 'Invoice': 'nunique'}).reset_index()
                          .sort_values(by = 'Revenue',ascending = False).head())

In [134]:
Top_revenue_customerId

Customer ID    Revenue  Quantity  Invoice
4157      18102.0  349164.35    124216       89
1621      14646.0  247210.30    170036       73
1257      14156.0  187634.50    107964       90
1821      14911.0  143290.94     69544      184
929       13694.0  129754.94    124800       94

In [135]:
dfcopy_sales.isnull().sum()


Invoice             0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
Price               0
Customer ID    102600
Country             0
Revenue             0
YearMonth           0
dtype: int64

In [136]:
(dfcopy_sales['Price'] <= 0).sum()

np.int64(0)

In [137]:
(dfcopy_sales['Quantity'] <= 0).sum()

np.int64(0)

In [138]:
dfcopy_sales.to_excel("clean_online_retail1.xlsx", index = False)

In [139]:
import csv

dfcopy_sales.to_csv(
    r"C:\Projects related to D.A\Project 1\clean_retail_1.csv",
    index=False,
    sep=",",
    quoting=csv.QUOTE_ALL,
    encoding="utf-8-sig"
)


In [140]:
dfcopy_sales.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate  Price  Customer ID         Country  Revenue YearMonth  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom     83.4   2009-12  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0   2009-12  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0   2009-12  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom    100.8   2009-12  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom     30.0   2009-12  
5 2009-12-01 07:45:00   1.65      13085.0  United Kingdom     39.6   2009-12  
6 2009-12-01 07:45:00   1.25      13085.0  United Kingdom     30.0   2009-12  
7 2009-12-01 07:45:00   5.95      13085.0  United Kingdom     59.5   2009-12  
8 2009-12-01 07:46:00   2.55      13085.0  United Kingdom     30.6   2009-12  
9 2009-12-01 07:46:00   3.75      13085.0  United Kingdom     45.0   2009-12

In [141]:
dfcopy_sales['Description'].astype(str).str.contains(',').sum()
# dfcopy_sales[dfcopy_sales.apply(lambda col:col.astype(str).str.contains(',')).any(axis=1)]

np.int64(10411)

In [142]:
dfcopy_sales[dfcopy_sales.apply(lambda col:col.astype(str).str.contains(',')).any(axis=1)]

Invoice StockCode                          Description  Quantity  \
9       489435     22349       DOG BOWL , CHASING BALL DESIGN        12   
91      489440     22349       DOG BOWL , CHASING BALL DESIGN         8   
107     489442    84251J         GREETING CARD, THE BLACK CAT        12   
144     489445     22241    GARLAND , WOODEN  " HAPPY EASTER"        24   
162     489446     22356     CHARLOTTE BAG , PINK/WHITE SPOTS        10   
...        ...       ...                                  ...       ...   
524148  538071    82613B       METAL SIGN,CUPCAKE SINGLE HOOK         1   
524149  538071    82613C       METAL SIGN,CUPCAKE SINGLE HOOK         1   
524164  538071    84247G             DECOUPAGE,GREETING CARD,         1   
524789  538148     21216     SET 3 RETROSPOT TEA,COFFEE,SUGAR         1   
525201  538156    17107D  FLOWER FAIRY,5 SUMMER B'DRAW LINERS        10   

               InvoiceDate  Price  Customer ID         Country  Revenue  \
9      2009-12-01 07:46:00   3.75      13085.0  United Kingdom    45.00   
91     2009-12-01 09:43:00   3.75      18087.0  United Kingdom    30.00   
107    2009-12-01 09:46:00   0.42      13635.0  United Kingdom     5.04   
144    2009-12-01 09:57:00   1.25      17519.0  United Kingdom    30.00   
162    2009-12-01 10:06:00   0.85      13758.0  United Kingdom     8.50   
...                    ...    ...          ...             ...      ...   
524148 2010-12-09 14:09:00   2.51          NaN  United Kingdom     2.51   
524149 2010-12-09 14:09:00   2.51          NaN  United Kingdom     2.51   
524164 2010-12-09 14:09:00   0.85          NaN  United Kingdom     0.85   
524789 2010-12-09 16:26:00  11.02          NaN  United Kingdom    11.02   
525201 2010-12-09 16:53:00   2.55      15555.0  United Kingdom    25.50   

       YearMonth  
9        2009-12  
91       2009-12  
107      2009-12  
144      2009-12  
162      2009-12  
...          ...  
524148   2010-12  
524149   2010-12  
524164   2010-12  
524789   2010-12  
525201   2010-12  

[10411 rows x 10 columns]

In [143]:
dfcopy_sales['Country'].astype(str).str.contains(',').sum()

np.int64(0)

## Revenue Analysis

#### Total Revenue:
9,785,571

#### Revenue by YearMonth:
   YearMonth     Revenue
1. 2010-11:     1426085.552
2. 2010-10:     1091642.650
3. 2010-09:     866815.201
4. 2009-12:     797577.130
5. 2010-03:     761444.501
   
#### Revenue by Country:
1. United Kingdom – Major contributor
2. EIRE
3. Netherlands

#### Top Products by Revenue:
- REGENCY CAKESTAND 3 TIER
- WHITE HANGING HEART T-LIGHT HOLDER
- ASSORTED COLOUR BIRD ORNAMENT

#### Key Observation:
Revenue is heavily dominated by UK market.

## Customer Analysis

####  Top Customer:
Customer ID 18102

- Total Invoices: 89
- Total Quantity: 124,216
- Total Revenue: 349,164

####  Revenue Concentration:
- Top 10 Customers contribute 14.59% of total revenue.
- Business is moderately diversified.